### PR TITLE
Adds context parallelism to FSDP2

### PR DESCRIPTION
**Pull request by jomitchellnv**
### Description

Creates a new training script callled `train_fsdp2_cp.py` where we add CP to FSDP2

#### Usage
This you can run this script in the same way that you execute `train_ddp_cp.py`

```python
torchrun --nproc_per_node=8 train_fsdp2_cp.py cp_size=<CP_SIZE>
```

For equivalence see
W B Chart 12_1_2025, 2_56_52 PM

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [X] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    logger.info("Initialized Model:\n%s", model)
+
+    # We call the transformer stack "layers" in our TE models, but it's called "layer" in the original ESM-2 models.
+    transformer_stack = model.esm.encoder.layers if hasattr(model.esm.encoder, "layers") else model.esm.encoder.layer
+    # Fully shard takes in a DeviceMesh object, which is a 2D mesh of dimensions (CP_dimension, DDP_dimension).
+    # FSDP2 will shard the model across the DDP (dim=1) dimension and then duplicate across the CP (dim=0) dimension.
+    for layer in transformer_stack:
+        fully_shard(layer, mesh=cp_dp_flat_mesh)
+        # Set CP group for layer if CP is enabled.
+        if args.cp_size > 1:
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py:123]: what happens if we set this and cp_size=1?

Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py:123]: I wouldn't set it if cp size is 1 since that would mean there's no CP?

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    # Validate that world_size is divisible by cp_size
+    if dist_config.world_size % args.cp_size != 0:
+        raise ValueError(
+            f"world_size ({dist_config.world_size}) must be divisible by cp_size ({args.cp_size}). "
+            f"Set cp_size to a divisor of world_size."
+        )
+
+    # Calculate DDP size (number of data parallel replicas)
+    ddp_size = dist_config.world_size // args.cp_size
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py]: dp_size?

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    
+
+    # Create a device mesh for DDP and CP.
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py]: DP yeah, not DDP

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        "cuda",
+        mesh_shape=(ddp_size, args.cp_size),
+        mesh_dim_names=("dp", "cp"),
+    )
+
+    # Our flattened group must have at least 2 ranks to enable Context Parallelism.
+    if ddp_size * args.cp_size <= 1:
+        cp_dp_mesh = device_mesh["dp", "cp"]._flatten(mesh_dim_name="dp_shard_cp")
+    else:
+        cp_dp_mesh = device_mesh
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py:83]: do we this this if/else because of pytorch/pytorch#169381?

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    transformer_stack = model.esm.encoder.layers if hasattr(model.esm.encoder, "layers") else model.esm.encoder.layer
+    # Fully shard takes in a DeviceMesh object, which is a 2D mesh of dimensions (CP_dimension, DDP_dimension).
+    # FSDP2 will shard the model across the DDP (dim=1) dimension and then duplicate across the CP (dim=0) dimension.
+    for layer in transformer_stack:
+        fully_shard(layer, mesh=cp_dp_mesh)
+        # Set CP group for layer if CP is enabled.
+        if args.cp_size > 1:
+            logger.debug(f"Rank {dist_config.rank}: Setting CP group for layer {layer}")
+            layer.set_context_parallel_group(
+                cp_group, torch.distributed.get_process_group_ranks(device_mesh["cp"].get_group()), torch.cuda.Stream()
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py]: `device_mesh["cp"].get_group()` is `cp_group`, right? But what is the input here for `torch.distributed.get_process_group_ranks(device_mesh["cp"].get_group())`? wouldn't we do something like `device_mesh.get_local_rank('cp')`?

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py (hunk 4 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py:83]: Yea -- and the else is so I can run this script with 1 GPU. I'll remove it when its fixed on their end

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py (hunk 5 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py]: >>device_mesh["cp"].get_group() is cp_group, right?
-----> yes precisely

torch.distributed.get_process_group_ranks(device_mesh["cp"].get_group()) returns a list of all global ranks that are part of this CP group. It needs global ranks not local ranks.

Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2_cp.py]: I updated to make it cleaner btw
